core/src/main/java/org/elasticsearch/action/DocumentRequest.java

```diff
@@ -70,4 +72,78 @@
      */
     String parent();
 
+    /**
+     * Get the document version for this request
+     * @return the document version
+     */
+    long version();
+
+    /**
+     * Sets the version, which will perform the operation only if a matching
+     * version exists and no changes happened on the doc since then.
+     */
+    T version(long version);
+
+    /**
+     * Get the document version type for this request
+     * @return the document version type
+     */
+    VersionType versionType();
+
+    /**
+     * Sets the versioning type. Defaults to {@link VersionType#INTERNAL}.
+     */
+    T versionType(VersionType versionType);
+
+    /**
+     * Get the requested document operation type of the request
+     * @return the operation type {@link OpType}
+     */
+    OpType opType();
+
+    /**
+     * Requested operation type to perform on the document
+     */
+    enum OpType {
+        /**
+         * Creates the resource. Simply adds it to the index, if there is an existing
+         * document with the id, then it won't be removed.
+         */
+        CREATE(0),
+        /**
+         * Index the source. If there an existing document with the id, it will
+         * be replaced.
+         */
+        INDEX(1),
+        /** Updates a document */
+        UPDATE(2),
+        /** Deletes a document */
+        DELETE(3);
+
+        private final byte op;
+        private final String lowercase;
+
+        OpType(int op) {
+            this.op = (byte) op;
+            this.lowercase = this.toString().toLowerCase(Locale.ENGLISH);
```

Comment:
changed it to ROOT, but there are a few places (e.g. `DocWriteResponse`) where we use ENGLISH too. maybe we should change it to ROOT as well?